src/rmp/src/abc_library_factory.cpp

```diff
@@ -0,0 +1,468 @@
+
+      if (!rise_gate_model || !fall_gate_model) {
+        logger_->error(utl::RMP,
+                       23,
+                       "rise/fall Gate model is empty for cell {}. Need timing "
+                       "information",
+                       cell->name());
+      }
+
+      AbcPopulateAbcSurfaceFromSta(&time_table->pCellRise,
+                                   rise_gate_model->delayModel());
+      AbcPopulateAbcSurfaceFromSta(&time_table->pCellFall,
+                                   fall_gate_model->delayModel());
+      AbcPopulateAbcSurfaceFromSta(&time_table->pRiseTrans,
+                                   rise_gate_model->slewModel());
+      AbcPopulateAbcSurfaceFromSta(&time_table->pFallTrans,
+                                   fall_gate_model->slewModel());
+    }
+
+    result.push_back(output_pin);
+  }
+
+  return result;
+}
+
+AbcLibraryFactory& AbcLibraryFactory::AddStaLibrary(
+    sta::LibertyLibrary* library)
+{
+  library_ = library;
+  return *this;
+}
+
+utl::deleted_unique_ptr<abc::SC_Lib> AbcLibraryFactory::Build()
+{
+  if (!library_) {
+    logger_->error(utl::RMP, 15, "Build called with null sta library");
+  }
+
+  abc::SC_Lib* abc_library = abc::Abc_SclLibAlloc();
+  PopulateAbcSclLibFromSta(abc_library);
+  return utl::deleted_unique_ptr<abc::SC_Lib>(
+      abc_library, [](abc::SC_Lib* lib) { abc::Abc_SclLibFree(lib); });
+}
+
+void AbcLibraryFactory::PopulateAbcSclLibFromSta(abc::SC_Lib* sc_library)
+{
+  sta::Units* units = library_->units();
+  sta::Unit* time_unit = units->timeUnit();
+  sta::Unit* cap_unit = units->capacitanceUnit();
+  sta::Unit* power_unit = units->powerUnit();
+
+  sc_library->pName = strdup(library_->name());
+  sc_library->pFileName = strdup(library_->filename());
+
+  sc_library->default_wire_load = nullptr;
+  sc_library->default_wire_load_sel = nullptr;
+  // TODO: Read and construct wireload tables
+
+  // unit selection
+
+  sc_library->unit_cap_fst = StaCapacitanceToAbc(cap_unit);
+  sc_library->unit_cap_snd
+      = ScaleAbbreviationToExponent(cap_unit->scaleAbbreviation());
+  sc_library->unit_time = StaTimeUnitToAbcInt(time_unit);
+
+  // defaults
+  float slew = -1.0;
+  bool max_slew_exists = false;
+  library_->defaultMaxSlew(slew, max_slew_exists);
+
+  if (max_slew_exists) {
+    sc_library->default_max_out_slew = time_unit->staToUser(slew);
+  } else {
+    sc_library->default_max_out_slew = -1.0;
+  }
+
+  // Loop through all of the cells in STA and create equivalents in
+  // the ABC structure.
+  sta::LibertyCellIterator cell_iterator(library_);
+  while (cell_iterator.hasNext()) {
+    sta::LibertyCell* cell = cell_iterator.next();
+
+    if (!IsCombinational(cell)) {
+      continue;
+    }
+
+    if (IsMultiOutputCell(cell)) {
+      continue;
+    }
+
+    if (!HasOutputs(cell)) {
+      continue;
+    }
+
+    if (HasNonInputOutputPorts(cell)) {
+      continue;
+    }
+
+    if (HasOnlyPowerAndGroundPorts(cell)) {
+      continue;
+    }
```

Comment:
Would it be simpler to just have IsAbcCompatible and traverse the pins once?

Reply:
I feel like that will end up being less clear overall since you need to traverse all the pins for some of the checks so you'll end up with a while loop with a bunch of counters that you then need to check. I feel like these checks could be moved into IsAbcCompatible, but I'd rather keep them seperate.

Reply:
If you have an output, how can you have only power and ground ports?

Reply:
They are considered inout ports

Reply:
inout are not outputs unless you mean to use isAnyOutput() not isOutput()

Reply:
Also they are not considered inout either but have their own categories.

Reply:
You're right. I removed it.